contracts/VolumeImbalanceRecorder.sol

```diff
@@ -0,0 +1,152 @@
+      if( uint(perBlockData.lastPriceUpdateBlock) != priceUpdateBlock ) continue;
+      if( perBlockData.lastBlock < latestBlock ) continue;
+
+      latestBlock = perBlockData.lastBlock;
+      buyImbalance = perBlockData.totalBuyUnitsImbalance;
+      if( uint(perBlockData.lastBlock) == currentBlock ) {
+        currentBlockImbalance = perBlockData.lastBlockBuyUnitsImbalance;
+      }
+    }
+
+    if( buyImbalance == 0 ) {
+      buyImbalance = getImbalanceInRange( token, priceUpdateBlock, currentBlock );
+    }
+  }
+
+  function getImbalance( ERC20 token, uint priceUpdateBlock, uint currentBlock)
+    public view returns(int totalImbalance, int currentBlockImbalance) {
+
+    int resolution = int(tokenControlInfo[token].minimalRecordResolution);
+
+    (totalImbalance,currentBlockImbalance) = getImbalanceSincePriceUpdate(token,
+                                                                          priceUpdateBlock,
+                                                                          currentBlock );
+    totalImbalance *= resolution;
+    currentBlockImbalance *= resolution;
+  }
+
+
+  function addImbalance( ERC20 token,
+                         int buyAmount,
+                         uint priceUpdateBlock,
+                         uint currentBlock ) internal {
+
+    uint currentBlockIndex = currentBlock % SLIDING_WINDOW_SIZE;
+    int64 recordedBuyAmount = int64(buyAmount /
+                         int(tokenControlInfo[token].minimalRecordResolution));
+
+    int prevImbalance = 0;
+
+    TokenImbalanceData memory currentBlockData = tokenImbalanceData[token][currentBlockIndex];
+    // first scenario - this is not the first tx in the current block
+    if( currentBlockData.lastBlock == currentBlock ) {
+      if( uint(currentBlockData.lastPriceUpdateBlock) == priceUpdateBlock ){
+        // just increase imbalance
+        currentBlockData.lastBlockBuyUnitsImbalance += recordedBuyAmount;
+        currentBlockData.totalBuyUnitsImbalance += recordedBuyAmount;
+      }
+      else {
+        // imbalance was changed in the middle of the block
+        prevImbalance = getImbalanceInRange( token, priceUpdateBlock, currentBlock );
+        currentBlockData.totalBuyUnitsImbalance = int64(prevImbalance) + recordedBuyAmount;
+        currentBlockData.lastBlockBuyUnitsImbalance += recordedBuyAmount;
+        currentBlockData.lastPriceUpdateBlock = uint64(priceUpdateBlock);
+      }
+    }
+    else {
+      // first tx in the current block
+      int currentBlockImbalance;
+      (prevImbalance, currentBlockImbalance) = getImbalanceSincePriceUpdate( token, priceUpdateBlock, currentBlock );
+
+      currentBlockData.lastBlockBuyUnitsImbalance = recordedBuyAmount;
+      currentBlockData.lastBlock = uint64(currentBlock);
+      currentBlockData.lastPriceUpdateBlock = uint64(priceUpdateBlock);
+      currentBlockData.totalBuyUnitsImbalance = int64(prevImbalance) + recordedBuyAmount;
```

Comment:
assuming we had a buy each block and that block starts from 1.
totalBuyUnitsImbalance for window index 1 will accumulate balance for blocks  , 1, 11, 21 etc.
or am I missing something?
I would expect in block 11 to have it accumulates all previous 10 (== sliding window size) blocks.

Reply:
you are missing something, the accumulation is not limited to the size of the sliding window. It could be infinite.
See how we use it [here](https://github.com/KyberNetwork/smart-contracts/blob/yaronvel/pricing/contracts/VolumeImbalanceRecorder.sol#L77)